File: bingo.js
// bingo.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

class BingoCard {
    constructor() {
        this.numbers = Array.from({length:5}, () => Array(5).fill(0));
        this.marked = Array.from({length:5}, () => Array(5).fill(false));
        this.generate();
    }

    generate() {
        const ranges = [[1,15],[16,30],[31,45],[46,60],[61,75]];
        for (let col=0; col<5; col++) {
            let nums = [];
            for (let i=ranges[col][0]; i<=ranges[col][1]; i++) nums.push(i);
            // перемешиваем
            for (let i=nums.length-1; i>0; i--) {
                const j = Math.floor(Math.random() * (i+1));
                [nums[i], nums[j]] = [nums[j], nums[i]];
            }
            for (let row=0; row<5; row++) {
                this.numbers[row][col] = nums[row];
            }
        }
        this.marked[2][2] = true; // FREE
    }

    mark(num) {
        for (let r=0; r<5; r++) {
            for (let c=0; c<5; c++) {
                if (this.numbers[r][c] === num) {
                    this.marked[r][c] = true;
                    return true;
                }
            }
        }
        return false;
    }

    checkWin() {
        // строки
        for (let r=0; r<5; r++) {
            if (this.marked[r].every(v => v)) return true;
        }
        // столбцы
        for (let c=0; c<5; c++) {
            let ok = true;
            for (let r=0; r<5; r++) if (!this.marked[r][c]) ok = false;
            if (ok) return true;
        }
        // диагонали
        let ok1 = true, ok2 = true;
        for (let i=0; i<5; i++) {
            if (!this.marked[i][i]) ok1 = false;
            if (!this.marked[i][4-i]) ok2 = false;
        }
        if (ok1 || ok2) return true;
        return false;
    }

    display(hide) {
        if (hide) {
            for (let r=0; r<5; r++) {
                let row = [];
                for (let c=0; c<5; c++) {
                    row.push(this.marked[r][c] ? colorize('XX', 'green') : '??');
                }
                console.log(row.join(' '));
            }
            return;
        }
        for (let r=0; r<5; r++) {
            let row = [];
            for (let c=0; c<5; c++) {
                if (this.marked[r][c]) {
                    row.push(colorize(String(this.numbers[r][c]).padStart(2), 'green'));
                } else {
                    row.push(String(this.numbers[r][c]).padStart(2));
                }
            }
            console.log(row.join(' '));
        }
    }
}

class BingoGame {
    constructor(mode) {
        this.mode = mode;
        this.players = [];
        this.called = [];
        this.current = 0;
        this.setupPlayers();
    }

    setupPlayers() {
        if (this.mode === 'single') {
            this.players = [
                { name: 'Игрок', card: new BingoCard(), won: false },
                { name: 'Компьютер', card: new BingoCard(), won: false }
            ];
        } else {
            this.players = [
                { name: 'Игрок 1', card: new BingoCard(), won: false },
                { name: 'Игрок 2', card: new BingoCard(), won: false }
            ];
        }
    }

    callNumber() {
        const available = [];
        for (let n=1; n<=75; n++) {
            if (!this.called.includes(n)) available.push(n);
        }
        if (available.length === 0) return -1;
        const idx = Math.floor(Math.random() * available.length);
        const num = available[idx];
        this.called.push(num);
        this.current = num;
        return num;
    }

    markAll(num) {
        for (const p of this.players) {
            p.card.mark(num);
        }
    }

    checkWinner() {
        for (let i=0; i<this.players.length; i++) {
            if (this.players[i].card.checkWin()) return i;
        }
        return -1;
    }

    displayState() {
        console.log(colorize('='.repeat(50), 'bold'));
        console.log(colorize(`Вызванное число: ${this.current || '—'}`, 'yellow'));
        for (let i=0; i<this.players.length; i++) {
            const p = this.players[i];
            console.log(colorize(`\n${p.name}:`, 'bold'));
            const hide = (this.mode === 'single' && i === 1);
            p.card.display(hide);
        }
        console.log(colorize('='.repeat(50), 'bold'));
    }

    async play() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const question = (q) => new Promise(resolve => rl.question(q, resolve));

        console.log(colorize('🎲 Добро пожаловать в Бинго!', 'bold'));
        if (this.mode === 'single') console.log('Игра против компьютера.');
        else console.log('Игра для двух игроков.');
        console.log('Нажимайте Enter для вызова следующего числа.');
        console.log('Для выхода введите q.\n');

        while (true) {
            this.displayState();
            const cmd = await question('> ');
            if (cmd.trim().toLowerCase() === 'q') {
                console.log('Выход.');
                rl.close();
                break;
            }
            if (cmd.trim() !== '') continue;
            const num = this.callNumber();
            if (num === -1) {
                console.log(colorize('Все числа вызваны! Ничья.', 'yellow'));
                break;
            }
            this.markAll(num);
            const winner = this.checkWinner();
            if (winner !== -1) {
                const name = this.players[winner].name;
                console.log(colorize(`🎉 Победитель: ${name}!`, 'green'));
                this.displayState();
                break;
            }
            await new Promise(r => setTimeout(r, 500));
        }
        rl.close();
    }
}

async function main() {
    let mode = 'single';
    let showStats = false, resetStats = false;
    const args = process.argv.slice(2);
    for (const arg of args) {
        if (arg === 'two') mode = 'two';
        else if (arg === '-s' || arg === '--stats') showStats = true;
        else if (arg === '-r' || arg === '--reset') resetStats = true;
        else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node bingo.js [single|two] [-s] [-r]');
            return;
        }
    }
    if (resetStats) {
        const f = path.join(os.homedir(), '.bingo_stats.json');
        if (fs.existsSync(f)) fs.unlinkSync(f);
        console.log('Статистика сброшена.');
        return;
    }
    if (showStats) {
        console.log('Статистика не реализована в JS для простоты.');
        return;
    }
    const game = new BingoGame(mode);
    await game.play();
}

main().catch(console.error);

File: bingo.go
// bingo.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type BingoCard struct {
	Numbers [5][5]int
	Marked  [5][5]bool
}

func NewBingoCard() *BingoCard {
	c := &BingoCard{}
	c.generate()
	return c
}

func (c *BingoCard) generate() {
	ranges := [5][2]int{{1, 15}, {16, 30}, {31, 45}, {46, 60}, {61, 75}}
	rand.Seed(time.Now().UnixNano())
	for col := 0; col < 5; col++ {
		nums := rand.Perm(ranges[col][1] - ranges[col][0] + 1)
		for row := 0; row < 5; row++ {
			c.Numbers[row][col] = ranges[col][0] + nums[row]
			c.Marked[row][col] = false
		}
	}
	c.Marked[2][2] = true // FREE
}

func (c *BingoCard) mark(num int) bool {
	for r := 0; r < 5; r++ {
		for col := 0; col < 5; col++ {
			if c.Numbers[r][col] == num {
				c.Marked[r][col] = true
				return true
			}
		}
	}
	return false
}

func (c *BingoCard) checkWin() (bool, string, int) {
	// Строки
	for r := 0; r < 5; r++ {
		ok := true
		for col := 0; col < 5; col++ {
			if !c.Marked[r][col] {
				ok = false
				break
			}
		}
		if ok {
			return true, "row", r
		}
	}
	// Столбцы
	for col := 0; col < 5; col++ {
		ok := true
		for r := 0; r < 5; r++ {
			if !c.Marked[r][col] {
				ok = false
				break
			}
		}
		if ok {
			return true, "col", col
		}
	}
	// Диагонали
	ok1, ok2 := true, true
	for i := 0; i < 5; i++ {
		if !c.Marked[i][i] {
			ok1 = false
		}
		if !c.Marked[i][4-i] {
			ok2 = false
		}
	}
	if ok1 {
		return true, "diag", 0
	}
	if ok2 {
		return true, "diag", 1
	}
	return false, "", 0
}

func (c *BingoCard) display(hide bool) {
	if hide {
		for r := 0; r < 5; r++ {
			for col := 0; col < 5; col++ {
				if c.Marked[r][col] {
					fmt.Print(colorize("XX", green) + " ")
				} else {
					fmt.Print("?? ")
				}
			}
			fmt.Println()
		}
		return
	}
	for r := 0; r < 5; r++ {
		for col := 0; col < 5; col++ {
			if c.Marked[r][col] {
				fmt.Printf("%s%2d%s ", colorize("", green), c.Numbers[r][col], colorize("", reset))
			} else {
				fmt.Printf("%2d ", c.Numbers[r][col])
			}
		}
		fmt.Println()
	}
}

type Player struct {
	Name string
	Card *BingoCard
	Won  bool
}

type BingoGame struct {
	Mode    string
	Players []Player
	Called  []int
	Current int
}

func NewBingoGame(mode string) *BingoGame {
	g := &BingoGame{Mode: mode}
	g.setupPlayers()
	return g
}

func (g *BingoGame) setupPlayers() {
	if g.Mode == "single" {
		g.Players = []Player{
			{"Игрок", NewBingoCard(), false},
			{"Компьютер", NewBingoCard(), false},
		}
	} else {
		g.Players = []Player{
			{"Игрок 1", NewBingoCard(), false},
			{"Игрок 2", NewBingoCard(), false},
		}
	}
}

func (g *BingoGame) callNumber() int {
	available := []int{}
	for n := 1; n <= 75; n++ {
		found := false
		for _, c := range g.Called {
			if c == n {
				found = true
				break
			}
		}
		if !found {
			available = append(available, n)
		}
	}
	if len(available) == 0 {
		return -1
	}
	idx := rand.Intn(len(available))
	num := available[idx]
	g.Called = append(g.Called, num)
	g.Current = num
	return num
}

func (g *BingoGame) markAll(num int) {
	for i := range g.Players {
		g.Players[i].Card.mark(num)
	}
}

func (g *BingoGame) checkWinner() int {
	for i, p := range g.Players {
		if win, _, _ := p.Card.checkWin(); win {
			return i
		}
	}
	return -1
}

func (g *BingoGame) displayState() {
	fmt.Println(colorize("==================================================", bold))
	if g.Current != 0 {
		fmt.Println(colorize(fmt.Sprintf("Вызванное число: %d", g.Current), yellow))
	} else {
		fmt.Println(colorize("Вызванное число: —", yellow))
	}
	for i, p := range g.Players {
		fmt.Println(colorize("\n"+p.Name+":", bold))
		hide := (g.Mode == "single" && i == 1)
		p.Card.display(hide)
	}
	fmt.Println(colorize("==================================================", bold))
}

func (g *BingoGame) play() {
	fmt.Println(colorize("🎲 Добро пожаловать в Бинго!", bold))
	if g.Mode == "single" {
		fmt.Println("Игра против компьютера.")
	} else {
		fmt.Println("Игра для двух игроков.")
	}
	fmt.Println("Нажимайте Enter для вызова следующего числа.")
	fmt.Println("Для выхода введите q.\n")

	scanner := bufio.NewScanner(os.Stdin)
	for {
		g.displayState()
		fmt.Print("> ")
		if !scanner.Scan() {
			break
		}
		cmd := strings.TrimSpace(scanner.Text())
		if cmd == "q" {
			fmt.Println("Выход.")
			break
		}
		if cmd != "" {
			continue
		}
		num := g.callNumber()
		if num == -1 {
			fmt.Println(colorize("Все числа вызваны! Ничья.", yellow))
			break
		}
		g.markAll(num)
		winner := g.checkWinner()
		if winner != -1 {
			name := g.Players[winner].Name
			fmt.Println(colorize(fmt.Sprintf("🎉 Победитель: %s!", name), green))
			g.displayState()
			break
		}
		time.Sleep(500 * time.Millisecond)
	}
}

func main() {
	mode := "single"
	showStats := false
	resetStats := false
	for _, arg := range os.Args[1:] {
		switch arg {
		case "two":
			mode = "two"
		case "-s", "--stats":
			showStats = true
		case "-r", "--reset":
			resetStats = true
		case "-h", "--help":
			fmt.Println("Usage: bingo [single|two] [-s] [-r]")
			return
		}
	}
	if resetStats {
		f := filepath.Join(os.Getenv("HOME"), ".bingo_stats.json")
		os.Remove(f)
		fmt.Println("Статистика сброшена.")
		return
	}
	if showStats {
		fmt.Println("Статистика не реализована в Go для простоты.")
		return
	}
	game := NewBingoGame(mode)
	game.play()
}
